File: leetcode/155-min-stack/Solution.java
class MinStack {
    public static class Node {
        public Node next;
        public Node minNext;
        public int val;
        public Node(int v) { this.val = v; }
    }
    private Node stackHead;
    private Node minHead;
    public MinStack() {}
    
    public void push(int val) {
        var node =  new Node(val);
        if(stackHead == null) {
            stackHead = node;
            minHead = node;
            return;
        }

        node.next = stackHead;
        stackHead = node;
        if(val < minHead.val) {
            node.minNext = minHead;
            minHead = node;
        }
    }
    
    public void pop() {
        if(stackHead == null) {
            return;
        }

        if(stackHead == minHead) {
            minHead = minHead.minNext != null ? minHead.minNext : null;
        }
        stackHead = stackHead.next != null ? stackHead.next : null;
    }
    
    public int top() {
        return stackHead != null ? stackHead.val : 0;
    }
    
    public int getMin() {
        return minHead != null ? minHead.val : 0;
    }
}

File: leetcode/155-min-stack/Solution.cpp
struct Node {
    int val;
    struct Node* next;
};
struct NodeRef {
    struct Node* ref;
    struct NodeRef* next;
};


class MinStack {
private:
    struct Node* head = NULL;
    struct NodeRef* min = NULL;
public:

    MinStack() {
    }
 
    void push(int val) {
        struct Node* nhead = new Node();
        nhead->val = val;
        nhead->next = head;
        head = nhead;     

        if(min == NULL){
            min = new NodeRef();
            min->ref = nhead;
        } else if(min->ref->val > val){
            struct NodeRef* nmin = new NodeRef();
            nmin->ref = nhead;
            nmin->next = min;
            min = nmin;
        }
    }
    
    void pop() {
        if(head == NULL) return;
        struct Node* ohead = head;
        head = head->next;

        if(min != NULL && min->ref == ohead) {
            struct NodeRef* omin = min;
            min = min->next;
            delete omin;
        }
        
        delete ohead;
    }
    
    int top() {
        return head != NULL ? head->val : 0;
    }
    
    int getMin() {
        return min != NULL ? min->ref->val : 0;
    }

    ~MinStack(){
        while(head != NULL){
            pop();
        }
    }   
};